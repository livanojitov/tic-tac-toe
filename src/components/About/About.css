.about{
  width: 95%;
  margin: 10px auto 0;
}

.about ul{
  margin-top: 0;
}

.about li {
  margin-top: 5px;
}