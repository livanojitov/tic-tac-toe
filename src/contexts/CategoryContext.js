import React, { createContext } from 'react';
import useFetchData from '../hooks/useFetchData';

export const CategoryContext = createContext();

const CategoryContextProvider = ({children}) => {
  let {data: categories, error} = useFetchData("categories");

  return (
    <CategoryContext.Provider value={{categories, error}}>
      {children}
    </CategoryContext.Provider>
  );

}
 
export default CategoryContextProvider;