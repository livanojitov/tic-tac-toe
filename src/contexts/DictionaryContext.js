import React, { createContext, useState, useEffect } from 'react';
import axios from 'axios'

export const DictionaryContext = createContext();

const DictionaryContextProvider = (props) => {
  const [DICTIONARY, setDictionary] = useState('');

  useEffect(() => {
    if (DICTIONARY.length === 0){
      axios.get('dictionary.json')
        .then(res => {
          setDictionary(res.data);
        })
    }
  });

  return (
    <DictionaryContext.Provider value={{ DICTIONARY }}>
      {props.children}
    </DictionaryContext.Provider>
  );

}
 
export default DictionaryContextProvider;