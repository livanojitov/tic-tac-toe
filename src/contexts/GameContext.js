import React, { createContext } from 'react';
import useLocalStorage          from '../hooks/useLocalStorage';
export const GameContext = createContext();

const GameContextProvider = ({children}) => {

  const {games, dispatch} = useLocalStorage('tictactoe', []);

  const getGame = (id) => {
    if (games.length > 0){
      return games.filter((game) => game.id === id)[0];
    } else {
      return false;
    }
  }

  return (
    <GameContext.Provider value={{games, dispatch, getGame}}>
      {children}
    </GameContext.Provider>
  );

}
 
export default GameContextProvider;