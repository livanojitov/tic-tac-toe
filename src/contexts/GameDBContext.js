import React, { createContext } from "react";
import useLocalStorage from "../hooks/useLocalStorage";
export const GameDBContext = createContext();

const GameDBContextProvider = ({ children }) => {
  const { games, dispatch } = useLocalStorage("tictactoe", []);

  const getGame = (id) => {
    if (games.length > 0) {
      return games.filter((game) => game.id === id)[0];
    } else {
      return false;
    }
  };

  return (
    <GameDBContext.Provider value={{ games, dispatch, getGame }}>
      {children}
    </GameDBContext.Provider>
  );
};

export default GameDBContextProvider;
