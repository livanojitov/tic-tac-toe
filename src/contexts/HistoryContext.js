import React, { createContext, useState } from 'react';

export const HistoryContext = createContext();

const HistoryContextProvider = (props) => {

  const [history, setHistory] = useState(0);

  const changeHistory = (history) => {
    setHistory(history);
  }

  return (
    <HistoryContext.Provider value={{ history, changeHistory }}>
      {props.children}
    </HistoryContext.Provider>
  );

}
 
export default HistoryContextProvider;