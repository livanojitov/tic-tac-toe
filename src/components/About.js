import React, { useState, useContext, useEffect } from 'react';
import { LanguageContext } from '../contexts/LanguageContext';
import { HistoryContext }  from '../contexts/HistoryContext';
import DICTIONARY          from './Dictionary';

const About = (props) => {
  const { getLanguage } = useContext(LanguageContext);
  const language = getLanguage();
  const { changeHistory } = useContext(HistoryContext);

  const [skills] = useState([
    { name: 'React.js',       url: 'https://reactjs.org/'                                                  },
    { name: 'React Hooks',    url: 'https://reactjs.org/docs/hooks-intro.html'                             },
    { name: 'Context API',    url: 'https://reactjs.org/docs/context.html'                                 },
    { name: 'Local Storage',  url: 'https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage'  },    
    { name: 'Javascript',     url: 'https://developer.mozilla.org/en-US/docs/Web/JavaScript'               },
    { name: 'ES6',            url: 'http://es6-features.org'                                               },
    { name: 'HTML5',          url: 'https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/HTML5'         },
    { name: 'CSS3',           url: 'https://www.w3.org/TR/selectors-3/'                                    }
  ]);
   
  const lis = skills.map((skill, key) => {
    return (
      <li key={key}><a target="_blank" href={skill.url} rel="noopener noreferrer">{skill.name}</a></li>
    )
  });

  const [source] = useState('https://github.com/lojito/tic-tac-toe');
  const [games]  = useState(
    { 
      name: 'Matching pairs ( built with Vue.js )', 
      url: 'https://lojito.github.io/matching-pairs'
    });

  useEffect(() => {
    return () => {
      if (props.history.location.pathname === '/history'){
        changeHistory(1);
      }else{
        changeHistory(0);
      }  
    }
  })

  return (

    <div className="about">

      <div>
        <span>{DICTIONARY[language].GAME_WRITTEN_WITH} </span>
        <ul>
          {lis}
        </ul>  
      </div> 

      <p>
          <span>{DICTIONARY[language].SOURCE_CODE} </span>
          <a target="_blank" href={source} rel="noopener noreferrer">{source}</a>
      </p>
      
      <span>{DICTIONARY[language].OTHER_GAMES} </span>            
      <ul>
        <li>
          <a target="_blank" href={games.url} rel="noopener noreferrer">{games.name}</a>
        </li>
      </ul>    
                  
    </div>

  )
} 

export default About;
