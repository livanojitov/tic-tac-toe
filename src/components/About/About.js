import React, { useContext, useEffect } from 'react';
import { DictionaryContext } from '../../contexts/DictionaryContext';
import './About.css'

const About = () => {

  const { dictionary, error } = useContext(DictionaryContext);  

  const techStack= [
    { name: 'React.js',                 url: 'https://reactjs.org/'                                                  },
    { name: 'React Hooks',              url: 'https://reactjs.org/docs/hooks-intro.html'                             },
    { name: 'Context API',              url: 'https://reactjs.org/docs/context.html'                                 },
    { name: 'Local Storage',            url: 'https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage'  },
    { name: 'Javascript',               url: 'https://developer.mozilla.org/en-US/docs/Web/JavaScript'               },
    { name: 'ES6',                      url: 'http://es6-features.org'                                               },
    { name: 'Firebase',                 url: 'https://firebase.google.com/docs'                                      },
    { name: 'HTML5',                    url: 'https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/HTML5'         },
    { name: 'CSS3',                     url: 'https://www.w3.org/TR/selectors-3/'                                    },
    { name: 'Jest',                     url: 'https://jestjs.io/'                                                    },
    { name: 'React Testing Library ',   url: 'https://testing-library.com/docs/react-testing-library/intro/'         }
  ];
  
  const techStackLi = techStack.map((tech, key) => {
    return (
      <li key={key}><a target="_blank" href={tech.url} rel="noopener noreferrer">{tech.name}</a></li>
    )
  });

  const source = 'https://github.com/lojito/tic-tac-toe';

  const games  = [
    { name: 'Puzzle (Angular)',                         url: 'https://github.com/lojito/puzzle'               },
    { name: 'Matching pairs (Vue.js)',                  url: 'https://github.com/lojito/matching-pairs'       },
    { name: 'Tic-tac-toe (Python)',                     url: 'https://github.com/lojito/python-tic-tac-toe'   },
    { name: 'Tic-tac-toe 4x4 (Django)',                 url: 'https://github.com/lojito/django-tictactoe'     },
    { name: 'UEFA Champions League (C)',                url: 'https://github.com/lojito/c_champions_league'   },
    { name: 'Bubble sort sorting algorithm (Assembly)', url: 'https://github.com/lojito/assembly-bubble-sort' }
  ];  

  const gamesLi = games.map((game, key) => {
    return (
      <li key={key}><a target="_blank" href={game.url} rel="noopener noreferrer">{game.name}</a></li>
    )
  });

  useEffect(() => {
    return () => {
      const chk = document.querySelector('input#menu');
      if (chk){
        chk.checked = false;
      }
    }
  });

  if (error){
    return <div className="error" data-testid="error">{error}</div>
  }

  return dictionary ? (
    <div className="about" data-testid="about">
      <div>
        <span data-testid="gameWrittenWith">{dictionary.GAME_WRITTEN_WITH}</span>
        <ul data-testid="techStack">
          {techStackLi}
        </ul>  
      </div> 
      <div>
        <span data-testid="sourceCode">{dictionary.SOURCE_CODE}</span>&nbsp;
        <ul data-testid="source">
          <li>
            <a target="_blank" href={source} rel="noopener noreferrer">{source}</a>
          </li>
        </ul>
      </div>
      <span data-testid="otherGames">{dictionary.OTHER_GAMES}</span>
      <ul data-testid="games">
        {gamesLi}
      </ul>
    </div>
  ) : (
    <div className="loading" data-testid="loading">Loading the About page...</div>
  )

} 

export default About;
