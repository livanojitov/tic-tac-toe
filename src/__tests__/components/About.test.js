import React from 'react';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { DictionaryContext } from '../../contexts/DictionaryContext';
import About from '../../components/About/About';

describe('About component', () => {
  let dictionary;
  let error;

  beforeEach(() => {
    dictionary = {GAME_WRITTEN_WITH : "This game has been developed with:",
                  SOURCE_CODE       : "Source Code:",
                  OTHER_GAMES       : "Other projects:"};
    error = '';
  });

  afterEach(cleanup);

  const renderAbout = () => {
    return render(
      <DictionaryContext.Provider value={{dictionary, error}}>
        <About/>
      </DictionaryContext.Provider>
    );
  }

  test("renders the About component correctly", () => {
    const { asFragment } = renderAbout();
    expect(asFragment()).toMatchSnapshot();
  })

  test("dictionary is not fetched yet", () => {
    dictionary = '';
    const { queryByTestId, getByTestId } = renderAbout();

    expect(getByTestId('loading')).toHaveTextContent('Loading the About page...');
    expect(queryByTestId('about')).toBeNull();
  })

  test("dictionary was fetched", () => {
    const { getByTestId } = renderAbout();
    expect(getByTestId('about')).not.toBeNull();
    expect(getByTestId('about')).toHaveClass('about');
  })

  test("error fetching the dictionary.json file", () => {
    dictionary = '';
    error = "Error while fetching the dictionary.json file from the Google Cloud";
    const { getByTestId } = renderAbout();

    expect(getByTestId('error')).toHaveTextContent(error);
  })

  test("GAME_WRITTEN_WITH, SOURCE_CODE and OTHER_GAMES descriptions", () => {
    const { getByTestId } = renderAbout();

    expect(getByTestId('gameWrittenWith')).toHaveTextContent(dictionary['GAME_WRITTEN_WITH']);
    expect(getByTestId('sourceCode')).toHaveTextContent(dictionary['SOURCE_CODE']);
    expect(getByTestId('otherGames')).toHaveTextContent(dictionary['OTHER_GAMES']);
  })

  test("techStack", () => {
    const { getByTestId } = renderAbout();

    expect(getByTestId('techStack').children.length).toBe(11);
  })

  test("source", () => {
    const { getByTestId } = renderAbout();

    expect(getByTestId('source')).toHaveTextContent('https://github.com/lojito/tic-tac-toe');
  })  

  test("games", () => {
    const { getByTestId } = renderAbout();

    expect(getByTestId('games').children.length).toBe(6);
  })  
})