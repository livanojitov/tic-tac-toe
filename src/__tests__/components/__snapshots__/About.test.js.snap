// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`About component renders the About component correctly 1`] = `
<DocumentFragment>
  <div
    class="about"
    data-testid="about"
  >
    <div>
      <span
        data-testid="gameWrittenWith"
      >
        This game has been developed with:
      </span>
      <ul
        data-testid="techStack"
      >
        <li>
          <a
            href="https://reactjs.org/"
            rel="noopener noreferrer"
            target="_blank"
          >
            React.js
          </a>
        </li>
        <li>
          <a
            href="https://reactjs.org/docs/hooks-intro.html"
            rel="noopener noreferrer"
            target="_blank"
          >
            React Hooks
          </a>
        </li>
        <li>
          <a
            href="https://reactjs.org/docs/context.html"
            rel="noopener noreferrer"
            target="_blank"
          >
            Context API
          </a>
        </li>
        <li>
          <a
            href="https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage"
            rel="noopener noreferrer"
            target="_blank"
          >
            Local Storage
          </a>
        </li>
        <li>
          <a
            href="https://developer.mozilla.org/en-US/docs/Web/JavaScript"
            rel="noopener noreferrer"
            target="_blank"
          >
            Javascript
          </a>
        </li>
        <li>
          <a
            href="http://es6-features.org"
            rel="noopener noreferrer"
            target="_blank"
          >
            ES6
          </a>
        </li>
        <li>
          <a
            href="https://firebase.google.com/docs"
            rel="noopener noreferrer"
            target="_blank"
          >
            Firebase
          </a>
        </li>
        <li>
          <a
            href="https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/HTML5"
            rel="noopener noreferrer"
            target="_blank"
          >
            HTML5
          </a>
        </li>
        <li>
          <a
            href="https://www.w3.org/TR/selectors-3/"
            rel="noopener noreferrer"
            target="_blank"
          >
            CSS3
          </a>
        </li>
        <li>
          <a
            href="https://jestjs.io/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Jest
          </a>
        </li>
        <li>
          <a
            href="https://testing-library.com/docs/react-testing-library/intro/"
            rel="noopener noreferrer"
            target="_blank"
          >
            React Testing Library 
          </a>
        </li>
      </ul>
    </div>
    <div>
      <span
        data-testid="sourceCode"
      >
        Source Code:
      </span>
      Â 
      <ul
        data-testid="source"
      >
        <li>
          <a
            href="https://github.com/lojito/tic-tac-toe"
            rel="noopener noreferrer"
            target="_blank"
          >
            https://github.com/lojito/tic-tac-toe
          </a>
        </li>
      </ul>
    </div>
    <span
      data-testid="otherGames"
    >
      Other projects:
    </span>
    <ul
      data-testid="games"
    >
      <li>
        <a
          href="https://github.com/lojito/puzzle"
          rel="noopener noreferrer"
          target="_blank"
        >
          Puzzle (Angular)
        </a>
      </li>
      <li>
        <a
          href="https://github.com/lojito/matching-pairs"
          rel="noopener noreferrer"
          target="_blank"
        >
          Matching pairs (Vue.js)
        </a>
      </li>
      <li>
        <a
          href="https://github.com/lojito/python-tic-tac-toe"
          rel="noopener noreferrer"
          target="_blank"
        >
          Tic-tac-toe (Python)
        </a>
      </li>
      <li>
        <a
          href="https://github.com/lojito/django-tictactoe"
          rel="noopener noreferrer"
          target="_blank"
        >
          Tic-tac-toe 4x4 (Django)
        </a>
      </li>
      <li>
        <a
          href="https://github.com/lojito/c_champions_league"
          rel="noopener noreferrer"
          target="_blank"
        >
          UEFA Champions League (C)
        </a>
      </li>
      <li>
        <a
          href="https://github.com/lojito/assembly-bubble-sort"
          rel="noopener noreferrer"
          target="_blank"
        >
          Bubble sort sorting algorithm (Assembly)
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
